blockchain/blockchain.go

```diff
@@ -240,9 +247,213 @@ func NewBlockChain(db database.DBManager, cacheConfig *CacheConfig, chainConfig
+
+	trieSync := state.NewStateSync(rootHash, targetDB.DiskDB())
+	maxBatchSize := 10000
+	queue := append([]common.Hash{}, trieSync.Missing(maxBatchSize)...)
+
+	committedCnt := 0
+
+	for trieSync.Pending() > 0 {
+		// TODO-Klaytn refine the status parameter (progress percentage, etc)
+		bc.committedCnt, bc.pendingCnt = committedCnt, trieSync.Pending()
+		logger.Warn("State migration progress", "committedCnt", bc.committedCnt, "pendingCnt", bc.pendingCnt)
+
+		results := make([]statedb.SyncResult, len(queue))
+
+		for i, hash := range queue {
+			data, err := srcCachedDB.NodeFromOld(hash)
+			if err != nil {
+				return fmt.Errorf("failed to retrieve node data for %x: %v", hash, err)
+			}
+
+			results[i] = statedb.SyncResult{Hash: hash, Data: data}
+		}
+
+		committedCnt += len(results)
+
+		if _, index, err := trieSync.Process(results); err != nil {
+			return fmt.Errorf("failed to process result #%d: %v", index, err)
+		}
+
+		if index, err := trieSync.Commit(targetDB.DiskDB().GetStateTrieMigrationDB()); err != nil {
+			return fmt.Errorf("failed to commit data #%d: %v", index, err)
+		}
+
+		queue = append(queue[:0], trieSync.Missing(maxBatchSize)...)
+
+		select {
+		case <-bc.stopStateMigration:
+			// TODO-Klaytn Revert DB.
+			// - copied new DB data to old DB.
+			// - remove new DB
+			return errors.New("stop state migration")
+		case <-bc.quit:
+			return nil
+		default:
+		}
+	}
+	logger.Info("completed to copy state migration", "resultCnt", committedCnt)
+
+	// Preimage Copy
+	// TODO-Klaytn consider to copy preimage
+
+	// Cross check that the two tries are in sync
+	// TODO-Klaytn consider to check Trie contents optionally
+	// TODO-Klaytn consider to check storage trie also
+	dirty, err := bc.checkTrieContents(targetDB, srcCachedDB, rootHash)
+	if err != nil || len(dirty) > 0 {
+		logger.Error("copied stated is invalid", "err", err, "len(dirty)", len(dirty))
+		// TODO-Klaytn Remove new DB and log.Error
+		if err != nil {
+			return err
+		}
+
+		return errors.New("copied state is not same with origin")
+	}
+
+	// TODO-Klaytn Swap DB and Remove Old DB
+	bc.db.FinishStateMigration()
+	logger.Info("completed state migration")
+
+	return nil
+}
+
+func (bc *BlockChain) checkTrieContents(oldDB, newDB *statedb.Database, root common.Hash) ([]common.Address, error) {
+	oldTrie, err := statedb.NewSecureTrie(root, oldDB)
+	if err != nil {
+		return nil, err
+	}
+
+	newTrie, err := statedb.NewSecureTrie(root, newDB)
+	if err != nil {
+		return nil, err
+	}
+
+	diff, _ := statedb.NewDifferenceIterator(oldTrie.NodeIterator([]byte{}), newTrie.NodeIterator([]byte{}))
+	iter := statedb.NewIterator(diff)
+
+	var dirty []common.Address
+
+	for iter.Next() {
+		key := newTrie.GetKey(iter.Key)
+		if key == nil {
+			return nil, fmt.Errorf("no preimage found for hash %x", iter.Key)
+		}
+
+		dirty = append(dirty, common.BytesToAddress(key))
+	}
+
+	return dirty, nil
+}
+
+func (bc *BlockChain) checkRestartStateMigration() {
```

Comment:
Why about changing this function name to `restartStateMigration()`?
We don't have any other restarting functions and checking `bc.db.InMigration` seems essential for restarting.

Reply:
reflected. Thanks. :)